basic-flatfield-correction/src/main.py

```diff
@@ -75,114 +124,481 @@ def _parse_files(fpath,regex,variables):
+    XAoffset = np.reshape(X_k_Aoffset,(options['size'],options['size']))
+    
+    # Update the weights
+    weight = 1/(np.abs(XE_norm) + options['epsilon'])
+    options['weight'] = weight * weight.size / np.sum(weight)
+    
+    # Calculate the mean flatfield and darkfield
+    temp = np.mean(XA,2) - XAoffset
+    flatfield = temp/np.mean(temp)
+    darkfield = XAoffset
+    return flatfield, darkfield, options
+
+def _get_photobleach(D,flatfield,darkfield=None):
+    D = np.reshape(D,(options['size']*options['size'],-1)).astype(np.float64)
+    if darkfield is None:
+        darkfield = np.zeros(flatfield.shape,dtype=np.float64)
+        
+    weights = np.ones(D.shape,dtype=np.float64)
+    epsilon = np.float64(0.1)
+    tol = np.float64(10**-6)
+    
+    for r in range(5):
+        W_idct_hat = np.reshape(flatfield,(-1,1))
+        A_offset = np.reshape(darkfield,(-1,1))
+        A1_coeff = np.reshape(np.mean(D,0),(1,-1))
+            
+        temp = np.linalg.svd(D,full_matrices=False,compute_uv=False)
+        norm_two = np.float64(temp[0])
+        mu = np.float64(12.5)/norm_two
+        mu_bar = mu * 10**7
+        rho = np.float64(1.5)
+        
+        d_norm = np.linalg.norm(D,'fro')
+        ent1 = 1
+        
+        A1_hat = np.zeros(D.shape,dtype=np.float64)
+        E1_hat = np.zeros(D.shape,dtype=np.float64)
+        Y1 = np.float64(0)
+        
+        iternum = 0
+        converged = False
+        
+        while not converged:
+            iternum += 1
+            A1_hat = np.matmul(W_idct_hat,A1_coeff) + A_offset
+            
+            E1_hat = E1_hat + np.divide(D - A1_hat - E1_hat + np.multiply(1/mu,Y1),ent1)
+            E1_hat = np.max(np.reshape(E1_hat - weights/(ent1*mu),(D.shape[0],D.shape[1],1)),-1,initial=0) + np.min(np.reshape(E1_hat + weights/(ent1*mu),(D.shape[0],D.shape[1],1)),-1,initial=0)
+            
+            R1 = D-E1_hat
+            
+            A1_coeff = np.reshape(np.mean(R1,0),(1, -1)) - np.mean(A_offset)
+            
+            A1_coeff[A1_coeff<0] = 0
+            
+            Z1 = D - A1_hat - E1_hat
+            
+            Y1 = Y1 + mu*Z1
+            
+            mu = np.min(mu*rho,initial=mu_bar)
+            
+            stopCriterion = np.linalg.norm(Z1,'fro')/d_norm
+            
+            if stopCriterion < tol:
+                converged = True
+                
+            logger.info('Iteration {}: stopCriterion {}'.format(iternum,stopCriterion))
+        
+        XE_norm = np.reshape(np.mean(A1_hat,0),(1,-1)) / E1_hat
+        weights = 1/np.abs(XE_norm + epsilon)
+        weights = weights * weights.size/np.sum(weights)
+        
+    return A1_coeff
 
 def main():
-    """ Initialize the logger """
-    logging.basicConfig(format='%(asctime)s - %(name)s - %(message)s', datefmt='%d-%b-%y %H:%M:%S')
-    logger = logging.getLogger(__name__)
-    logger.setLevel(logging.INFO)
+    import bioformats
+    import javabridge as jutil
```

Comment:
Why do you bring these in here? Why not outside the def?

Reply:
Mainly to make it apparent that javabridge/bioformats has to be initialized by whoever is using these tools. They are not handled under the hood for reasons explained in the bfio tool.